feat : 알림 기능 구현

NCookies opened the pull request:
## 작업 내용

- 알림 이벤트 발행 및 처리
- 알림 데이터 저장 (NoSQL 사용)
- 해당 유저에게 알림 발송 (웹소켓 사용)
- 알림 목록 조회
    - 알림 이벤트 발생 시 연결되어 있지 않았던 유저가 GET 요청하면 웹소켓으로 발송
- 알림 읽음 처리

---

## 변경 사항

- 도메인 계층에서 알림(notification) 제거
- redis 관련 설정

---

## 트러블 슈팅

- 알림을 현재 프로젝트 아키텍처에 어떻게 적용할지 ([5분 기록 보드](https://www.notion.so/teamsparta/1-5-2002dc3ef51481e7afbec86e90d0010e?p=2092dc3ef5148070bf76f4d8b57302c8&pm=s))

---

## 해결해야 할 문제

- 어떤 DB 사용할지 의논 필요
  - 현재는 임시로 redis 사용 중이나 추후 mongoDB 등으로 대체될 수 있음
  - redis는 휘발성이므로...


---

## 참고 사항

- MBA-37

---

### 코드 리뷰 전 확인 체크리스트

- [x]  불필요한 콘솔 로그, 주석 제거
- [x]  커밋 메시지 컨벤션 준수 (`type : `)
- [x]  기능 정상 동작 확인


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **신규 기능**
  - 커뮤니티 활동(댓글 작성, 추천 등)에 따른 알림 생성 및 발송 기능이 강화되었습니다.
  - 알림 목록 조회 및 읽음 처리 API가 제공됩니다.
  - 실시간 웹소켓(STOMP) 기반 알림 수신 기능이 추가되었습니다.
- **기타**
  - 알림 데이터가 Redis 기반으로 저장 및 관리됩니다.
  - 채팅 페이지에서 실시간 알림 메시지를 콘솔에서 확인할 수 있습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

NCookies committed:
feat : .gitkeep 파일 제거

NCookies committed:
feat : 도메인에서 notification 제거

- 알림 기능은 자체적인 비즈니스 로직을 가지고 있지 않음
  - 알림 저장, 발송, 읽음 처리라는 간단한 기능만 필요
  - 때문에 이를 완전히 인프라 계층에서만 구현해도 문제가 없다고 판단했다.
- 알림 데이터 저장은 nosql 사용하기로 함
  - 다양한 형태의 데이터 가질 수 있기 때문
  - 우선은 임시로 redis 사용
  - 추후 mongoDB 등으로 대체될 수 있음

NCookies committed:
feat : 알림을 redis에 저장하고 조회, 읽음 처리하는 기능 구현

NCookies committed:
feat : 알림 이벤트 처리 및 발송 기능 구현

NCookies committed:
feat : 댓글, 추천 시 해당하는 유저에게 알림 발송되도록 구현

thezz9 commented on src/main/java/org/ezcode/codetest/application/community/service/ReplyService.java:
```diff
@@ -111,4 +123,19 @@ public void removeReply(Long problemId, Long discussionId, Long replyId, Long us
+
+	private void notify(User sender, User recipient, Reply reply) {
+
+		if (recipient == null || sender.getId().equals(recipient.getId())) {
```
이 부분 User 안에 도메인 메서드로 추가하시는 건 어떨까용?

NCookies replied:
좋은 것 같습니다!

thezz9 commented on src/main/java/org/ezcode/codetest/application/notification/enums/NotificationType.java:
```diff
@@ -0,0 +1,31 @@
+		return Arrays.stream(NotificationType.values())
+			.filter(n -> n.name().equalsIgnoreCase(notificationType))
+			.findFirst()
+			.orElseThrow(() -> new IllegalArgumentException("Invalid input : " + notificationType));
```
요 부분은 나중에 CustomException으로 대체하시는 건가요??

NCookies replied:
개발하면서 사용 안 하게 된 코드라서 삭제하려고 합니다.

다만 enum도 이렇게 from으로 파싱하도록 구현되어 있던데 enum 전용 예외 클래스를 만들지, 도메인에 해당하는 예외를 사용할지는 의논해 보면 좋을 것 같습니다.

thezz9 commented on src/main/java/org/ezcode/codetest/infrastructure/event/dto/NotificationResponse.java:
```diff
@@ -0,0 +1,41 @@
+
+) {
+
+	public static NotificationResponse from(NotificationRecord record) {
+		return NotificationResponse
+			.builder()
+			.id(record.getId())
+			.type(record.getType())
+			.message(record.getMessage())
+			.redirectUrl(record.getRedirectUrl())
+			.payload(record.getPayload())
+			.isRead(record.isRead())
+			.createdAt(record.getCreatedAt())
+			.build();
+	}
```
레코드 클래스를 사용하면 내부적으로 모든 파라미터를 받는 생성자가 생성된다고 합니다.
그래서 모든 파라미터를 순서대로 받는 스태틱 메서드가 있으면 리턴을 빌더보단 new로 하는 게 좋다고 해요
둘이 지향점이 다르다고 합니다. 빌더는 유연하고 복잡한 객체 생성, 레코드는 간단하고 명확한 불변 객체
저도 그래서 이번 PR부터는 record + new 조합이나 class + builder 조합으로 바꿨어요

NCookies replied:
의견 감사합니다! 파라미터 개수가 많아져서 builder가 더 좋다고 생각했는데 무조건 그런건 아닌가 보군요.

NCookies committed:
feat : PR 피드백 반영

- 유저 비교를 entity 메서드에서 처리
- 사용되지 않는 메서드 삭제
  - enum에서 공통적으로 사용하는 메서드 안에서 예외를 어떻게 던질지 의논 필요
- record에서 builder 대신 new 생성자 사용

NCookies committed:
feat : 알림 발송 조건 로직 수정
